### PR TITLE
Implement unique

### PR DESCRIPTION
**Pull request by not522**
I implement `cupy.unique`. It is requested in #813.
This PR does not support `axis` parameter which needs sort of consolidated array.

### CODE REVIEW

File: cupy/manipulation/add_remove.py [hunk 1]
```diff
@@ -16,4 +16,74 @@
...
+    ret = aux[mask],
+    if return_index:
+        ret += perm[mask],
+    if return_inverse:
+        imask = cupy.cumsum(mask) - 1
+        inv_idx = cupy.empty(mask.shape, dtype=cupy.intp)
+        inv_idx[perm] = imask
+        ret += inv_idx,
+    if return_counts:
+        idx = cupy.concatenate(cupy.nonzero(mask) + (cupy.array([mask.size]),))
```
Comment (okuta) [inline cupy/manipulation/add_remove.py]: How about following code?
```
nonzero = cupy.nonzero(mask)
idx = cupy.array((nonezero.size + 1,), nonzero.dtype)
idx[:-1] = nonzero
idx[-1] = mask.size
```

File: cupy/manipulation/add_remove.py [hunk 2]
```diff
@@ -16,4 +16,74 @@
...
+        perm = ar.argsort()
+        aux = ar[perm]
+    else:
+        ar.sort()
+        aux = ar
+    mask = cupy.empty(aux.shape, dtype=cupy.bool_)
+    mask[0] = True
+    mask[1:] = aux[1:] != aux[:-1]
+
+    ret = aux[mask],
```
Comment (okuta) [inline cupy/manipulation/add_remove.py]: How about following code?
```
ret = aux[mask]
if return_index or return_inverse or return_counts:
        return ret
ret  = ret,
...
...

return ret

File: cupy/manipulation/add_remove.py [hunk 3]
```diff
@@ -16,4 +16,74 @@
...
+              unique array. Only provided if `return_inverse` is True.
+            * The number of times each of the unique values comes up in the
+              original array. Only provided if `return_counts` is True.
+
+    .. seealso:: :func:`numpy.unique`
+    """
+    if axis is not None:
+        raise NotImplementedError('axis option is not supported yet.')
+
+    ar = ar.flatten()
```
Comment (okuta) [inline cupy/manipulation/add_remove.py]: Please use `cupy.asarray` to support array-like object.

**Comment by okuta** [comment]
> I think it is good to implement the standard way to test them.

I think so too. Do you have any idea?

**Comment by not522** [comment]
Let's move to #1206 and discuss.

**Comment by okuta** [comment]
OK, Let's discuss on that issue. Can I merge this PR tentatively?

**Comment by not522** [comment]
I think OK to merge if you don't have other concerns.

**Comment by okuta** [comment]
LGTM!

**Comment by MarwanAbdelatti** [comment]
I came across the discussion here but I couldn't find how the (axis) option is supported. I'm trying to get unique **_rows_** from a cupy array.

**Pull request merged**